### PR TITLE
wait 40 seconds after a api call in order to make it easier to face i…

### PR DESCRIPTION
**Pull request by bladewang**
#54 

 * refactory `download_books`: to download a batch of books as soon as you get them
 * add option `--wait` to setup wait seconds, default 40 secs
 * add option `--batch-size` to setup batch size in a api call
 * update readme

### CODE REVIEW

**Comment by yihong0618** [comment]
突然想，既然我们每次等 40s.
那么我们可以修改现在的逻辑，一个 batch 一个 batch 的下，而不是一起拿到所有书籍 asins?

**Comment by bladewang** [comment]
> 

这次的修改就是每个 batch 拿回来就直接下载的哈

File: kindle.py [hunk 1]
```diff
@@ -290,19 +294,24 @@ def download_one_book(self, book, device, index, filetype="EBOK"):
...
-        for book in books:
-            self.download_one_book(book, device, index, filetype)
-            index += 1
+        while True:
+            result = self.get_one_batch_books(filetype=filetype, start_index=index, batchsize=batchsize)
+            books = result["OwnershipData"]["items"]
+            for book in books:
+                self.download_one_book(book, device, index, filetype)
+                index += 1
+                sleep(wait_secs)
```
Comment (yihong0618) [inline kindle.py:310]: how this loop break?

File: kindle.py [hunk 2]
```diff
@@ -377,6 +386,20 @@ def download_books(self, start_index=0, filetype="EBOK"):
         default="EBOK",
         help="to download personal documents or ebook",
     )
+    parser.add_argument(
+        "--batch-size",
+        dest="batchsize",
+        type=int,
+        default=18,
```
Comment (yihong0618) [inline kindle.py:393]: ebooks keep batch-size is fine we tested

File: kindle.py [hunk 3]
```diff
@@ -173,74 +174,77 @@ def get_devices(self):
...
+        TODO: refactor this function
+        """
+        startIndex = start_index
+        books = []
+        while True:
+            try:
+                result = self.get_one_batch_books(startIndex, filetype, batchsize)
+            except ValueError as e:
+                if e.args[0] == 503:
+                    break
```
Comment (yihong0618) [inline kindle.py:240]: if not we log the `e`

File: kindle.py [hunk 4]
```diff
@@ -173,74 +174,77 @@ def get_devices(self):
...
+            books.extend(result["OwnershipData"]["items"])
             if result["OwnershipData"]["hasMoreItems"]:
-                startIndex += batchSize
-                payload["param"]["OwnershipData"]["startIndex"] = startIndex
+                startIndex += batchsize
             else:
                 break
+            print(f"{startIndex} / {self.total_to_download}")
```
Comment (yihong0618) [inline kindle.py:246]: more info will be nicer?

**Review by yihong0618 — CHANGES_REQUESTED** [review]
我感觉我们 ebook 还是保存 100 比较好毕竟不用 sleep 
pdoc 的话 18.

**Comment by yihong0618** [comment]
In my env still 503

**Comment by yihong0618** [comment]
Maybe we can choose another idea.

File: kindle.py (hunk 1 above)
Comment (bladewang) [reply kindle.py:310]: My bad, this loop won't quit.
I'll close this PR and submit it later

Comment (bladewang) [reply kindle.py:310]: > 我感觉我们 ebook 还是保存 100 比较好毕竟不用 sleep 
> pdoc 的话 18.

OK，我修改一下再做提交

**Comment by bladewang** [comment]
> Maybe we can choose another idea.

That's true, more ideas will be better.

**Pull request closed without merging**